cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/StateVariablesExport.java

```diff
@@ -125,19 +131,161 @@ private static void buildAngleRefs(Bus bus, Map<String, String> angleRefs, Cgmes
+        //      following string values: “converged” and “diverged” which represents the
+        //      convergence status of the cim:TopologicalIsland.
+        // Consider the island is converged unless we add a non-converged bus
+        String loadFlowStatus = CONVERGED;
+        final double maxPMismatchConverged;
+        final double maxQMismatchConverged;
+        final Set<Bus> checkedBusViewBuses = new HashSet<>();
+        final boolean checkConvergedInAllBuses;
+
+        private TopologicalIsland(String key, List<String> topologicalNodes, CgmesExportContext context) {
+            this.key = key;
+            this.checkConvergedInAllBuses = false;
+            this.topologicalNodes = topologicalNodes;
+            this.maxPMismatchConverged = context.getMaxPMismatchConverged();
+            this.maxQMismatchConverged = context.getMaxQMismatchConverged();
+        }
+
+        static TopologicalIsland fromSynchronousComponent(String key, CgmesExportContext context) {
+            return new TopologicalIsland(key, new ArrayList<>(), context);
+        }
+
+        static TopologicalIsland fromTopologicalNode(String topologicalNode, CgmesExportContext context) {
+            return new TopologicalIsland(topologicalNode, Collections.singletonList(topologicalNode), context);
+        }
+
+        void addNode(String topologicalNode, Bus bus, boolean updateLoadFlowStatus) {
+            topologicalNodes.add(topologicalNode);
+            if (updateLoadFlowStatus) {
+                updateLoadFlowStatus(bus);
+            }
+        }
+
+        void updateLoadFlowStatus(Bus bus) {
+            if (loadFlowStatus.equals(DIVERGED) && !checkConvergedInAllBuses) {
+                return;
+            }
+            if (!(isValidVoltage(bus.getV()) && isValidAngle(bus.getAngle()) && isInAccordanceWithKirchhoffsFirstLaw(bus))) {
+                loadFlowStatus = DIVERGED;
+            }
+        }
+
+        boolean isValidVoltage(double v) {
+            return v >= 0.01;
+        }
+
+        boolean isValidAngle(double a) {
+            return Double.isFinite(a);
+        }
+
+        boolean isInAccordanceWithKirchhoffsFirstLaw(Bus bus) {
+            // Instead of checking switch flows, we check that the corresponding bus view bus is balanced
+            Optional<Bus> optionalBusViewBus = getBusViewBus(bus);
+            if (optionalBusViewBus.isEmpty()) {
+                LOG.error("Can not check if bus is in accordance with Kirchhoff's first law. No BusView bus can be found for: {}", bus);
+                return false;
+            }
+            Bus busViewBus = optionalBusViewBus.get();
+            // We do not check the same bus view bus more than once
+            if (checkedBusViewBuses.contains(busViewBus)
+                    || busViewBus.getConnectedTerminalCount() == 0
+                    || isSlack(busViewBus)
+                    || isAngleReference(busViewBus)) {
```

Comment:
Why do you have to filter the angle reference bus?

Reply:
Right. Fixed. Buses for angle reference terminals must also be solved, they should not have a mismatch.